R/geom-label.R

```diff
@@ -121,32 +122,41 @@ labelGrob <- function(label, x = unit(0.5, "npc"), y = unit(0.5, "npc"),
   if (!is.unit(y))
     y <- unit(y, default.units)
 
-  gTree(label = label, x = x, y = y, just = just, padding = padding, r = r,
-    name = name, text.gp = text.gp, rect.gp = rect.gp, vp = vp, cl = "labelgrob")
-}
+  if (!is.null(angle) & is.null(vp)) {
+    vp <- viewport(
+      angle = angle, x = x, y = y,
+      width = unit(0, "cm"), height = unit(0, "cm"),
+      gp = gpar(fontsize = text.gp$fontsize)
+    )
+    x <- unit(rep(0.5, length(x)), "npc")
+    y <- unit(rep(0.5, length(y)), "npc")
+  }
 
-#' @export
-makeContent.labelgrob <- function(x) {
-  hj <- resolveHJust(x$just, NULL)
-  vj <- resolveVJust(x$just, NULL)
-
-  t <- textGrob(
-    x$label,
-    x$x + 2 * (0.5 - hj) * x$padding,
-    x$y + 2 * (0.5 - vj) * x$padding,
-    just = c(hj, vj),
-    gp = x$text.gp,
-    name = "text"
+  descent <- font_descent(
+    text.gp$fontfamily, text.gp$fontface, text.gp$fontsize, text.gp$cex
+  )
+  hjust <- resolveHJust(just, NULL)
+  vjust <- resolveVJust(just, NULL)
+
+  if (!inherits(padding, "margin")) {
+    padding <- rep(padding, length.out = 4)
+  }
+
+  text <- titleGrob(
+    label = label, hjust = hjust, vjust = vjust,
+    x = x + (0.5 - hjust) * sum(padding[c(2, 4)]),
+    y = y + (0.5 - vjust) * sum(padding[c(1, 3)]),
+    margin = padding, margin_x = TRUE, margin_y = TRUE,
+    gp = text.gp
   )
 
-  r <- roundrectGrob(x$x, x$y, default.units = "native",
-    width = grobWidth(t) + 2 * x$padding,
-    height = grobHeight(t) + 2 * x$padding,
-    just = c(hj, vj),
-    r = x$r,
-    gp = x$rect.gp,
-    name = "box"
+  box <- roundrectGrob(
+    x = x, y = y - 0.5 * descent,
+    width  = grobWidth(text),
+    height = grobHeight(text),
+    just   = c(hjust, vjust),
+    r = r, gp = rect.gp, name = "box"
   )
 
-  setChildren(x, gList(r, t))
+  gTree(children = gList(box, text), name = name, vp = vp)
```

Comment:
Do you have a reason for this change. AFAIK, the `makeContent()` method expects you to return a modified version of the input, not a new object (i.e. use `setChildren()`). This is because grid keeps track of the actual objects under the hood

Reply:
Yeah the reason is that we don't need `makeContent()` here, so I've included all necessary steps in the `labelGrob()` constructor. This simplifies the grob a lot, and the `makeContent()` method wasn't doing any calculations that rely on the device size or resolution.

Reply:
Ah - I got confused by the diff... this is no longer a `makeContent` method

Reply:
Yep that's right :)